packages/web/src/lib/core/auth/internal/StitchAuthImpl.ts

```diff
@@ -125,4 +277,102 @@ export default class StitchAuthImpl extends CoreStitchAuth<StitchUser>
       });
     }
   }
+
+  // The state we generate is to be used for any kind of request where we will
+  // complete an authentication flow via a redirect. We store the generate in
+  // a local storage bound to the app's origin. This ensures that any time we
+  // receive a redirect, there must be a state parameter and it must match
+  // what we ourselves have generated. This state MUST only be sent to
+  // a trusted Stitch endpoint in order to preserve its integrity. Stitch will
+  // store it in some way on its origin (currently a cookie stored on this client)
+  // and use that state at the end of an auth flow as a parameter in the redirect URI.
+  private generateState() {
+    let state = "";
+    for (let i = 0; i < 64; ++i) {
+      state += alpha.charAt(Math.floor(Math.random() * alpha.length));
+    }
+
+    return state;
+  }
+
+  private pageRootUrl() {
+    return [
+      window.location.protocol,
+      "//",
+      window.location.host,
+      window.location.pathname
+    ].join("");
+  }
+
+  private prepareRedirect(
+    credential: StitchRedirectCredential,
+    redirectType: RedirectTypes
+  ): { redirectUrl: string; state: string } {
+    this.storage.set(RedirectKeys.RedirectProvider, credential.providerType);
+    let redirectUrl = credential.redirectUrl;
+    if (redirectUrl === undefined) {
+      redirectUrl = this.pageRootUrl();
+    }
+
+    const state = this.generateState();
+    this.storage.set(RedirectKeys.State, state);
+    this.storage.set(RedirectKeys.Type, redirectType);
+
+    return { redirectUrl, state };
+  }
+
+  private unmarshallUserAuth(data): AuthInfo {
+    const parts = data.split("$");
+    if (parts.length !== 4) {
+      throw new RangeError("invalid user auth data provided: " + data);
+    }
+
+    return new AuthInfo(parts[2], parts[3], parts[0], parts[1]);
```

Comment:
[nit] it would be nice if the parts were labeled so this is a little more readable